subprojects/core/src/main/java/org/gradle/api/AttributeValue.java

```diff
@@ -0,0 +1,252 @@
+        return new AttributeValue<T>(value);
+    }
+
+    /**
+     * Creates a missing attribute value, used to represent the fact that the attribute is known
+     * but the consumer didn't want to express a value for it (it doesn't care).
+     * @param <T> the type of the attribute
+     * @return a <i>missing</i> attribute value
+     */
+    public static <T> AttributeValue<T> missing() {
+        return Cast.uncheckedCast(MISSING);
+    }
+
+    /**
+     * Creates an unknown attribute value, used to represent the fact that the attribute is unknown
+     * from the consumer side. Produces might want to do something differently in this case.
+     * @param <T> the type of the attribute
+     * @return an <i>unknown</i> attribute value
+     */
+    public static <T> AttributeValue<T> unknown() {
+        return Cast.uncheckedCast(UNKNOWN);
+    }
+
+    /**
+     * Tells if this attribute value is present.
+     * @return true if this attribute value is present, implying not <code>null</code>.
+     */
+    public boolean isPresent() {
+        return value != null;
+    }
+
+    /**
+     * Returns the value of this attribute.
+     * @return the value of this attribute. Throws an error if called on a missing or unknown attribute value.
+     */
+    public T get() {
+        if (value == null) {
+            throw new NoSuchElementException("No value provided");
+        }
+        return value;
+    }
+
+    /**
+     * Returns true if this attribute value is missing.
+     * @return true if this attribute value is missing.
+     */
+    public boolean isUnknown() {
+        return false;
+    }
+
+    /**
+     * Returns true if this attribute value is unknown.
+     * @return true if this attribute value is unknown.
+     */
+    public boolean isMissing() {
+        return false;
+    }
+
+    /**
+     * Configures a value to return if this attribute value is present. The transformer instance is passed
+     * the value of the attribute when called. This can be used to return a different value depending on
+     * the state of the attribute, as in this example:
+     * <p/>
+     * <code>
+     *     String result = attributeValue.whenPresent { String value ->
+     *         value.toUpperCase()
+     *     }.whenMissing {
+     *         'default'
+     *     }.whenUnknown {
+     *         'unknown
+     *     }.get()
+     * </code>
+     * <p/>
+     * Alternatively if the missing and default value are the same, it is possible to use the shortcut {@link DefaultingResultBuilder#getOrElse} method:
+     * <p/>
+     * <code>
+     *     String result = attributeValue.whenPresent { String value ->
+     *         value.toUpperCase()
+     *     }.getOrElse {
+     *         'default'
+     *     }
+     * </code>
+     * @param transformer the transformer
+     * @param <O> the type of the result
+     * @return a value producer
+     */
+    public <O> ResultBuilder<O> whenPresent(Transformer<O, T> transformer) {
```

Comment:
It does add a bit of API surface, but I think it is valuable. It is verbose because of Java, basically. It allows the user to start with an {{AttributeValue}} and tell what to do in various cases, then get the result. But it prevents the user from generating incomplete implementations. Basically, you can do:

```
value.whenPresent { ... }
        .whenMissing { ... }
        .whenUnknown { ... }
        .get()
```

or

```
value.whenPresent { ... }
        .whenMissing { ... }
        .getOrElse  { ... }
```

but this wouldn't compile:

```
value.whenPresent { ... }
        .get()
```

It's basically compile-time safety.